src/sentry/api/endpoints/source_map_debug.py

```diff
@@ -185,15 +170,130 @@ def _get_filename_and_path(self, exception, frame_idx):
+
+    def _verify_dist_matches(self, release, event, artifact, filename):
+        try:
+            if event.dist is None and artifact.dist_id is None:
+                return
+            dist = Distribution.objects.get(release=release, name=event.dist)
+        except Distribution.DoesNotExist:
+            raise Exception(
+                SourceMapProcessingIssue.DIST_MISMATCH,
+                {"eventDist": event.dist, "filename": filename},
+            )
+        if artifact.dist_id != dist.id:
+            raise Exception(
+                SourceMapProcessingIssue.DIST_MISMATCH,
+                {"eventDist": dist.id, "artifactDist": artifact.dist_id, "filename": filename},
+            )
+        return True
+
+    def _find_matching_artifact(
+        self, release_artifacts, urlparts, abs_path, filename, release, event
+    ):
+        unified_path = self._unify_url(urlparts)
+        full_matches, partial_matches = self._find_matches(
+            release_artifacts, unified_path, filename, release, event
+        )
+
+        artifact_names = [artifact.name for artifact in release_artifacts]
+
+        if len(full_matches) == 0:
+            if len(partial_matches) > 0:
+                partial_match = partial_matches[0]
+                url_prefix = self._find_url_prefix(filename, partial_match.name)
+
+                raise Exception(
+                    SourceMapProcessingIssue.PARTIAL_MATCH,
+                    {
+                        "absPath": abs_path,
+                        "partialMatchPath": partial_match.name,
+                        "filename": filename,
+                        "unifiedPath": unified_path,
+                        "urlPrefix": url_prefix,
+                        "artifactNames": artifact_names,
+                    },
+                )
+
+            raise Exception(
+                SourceMapProcessingIssue.NO_URL_MATCH,
+                {
+                    "absPath": abs_path,
+                    "filename": filename,
+                    "unifiedPath": unified_path,
+                    "artifactNames": artifact_names,
+                },
+            )
+        return full_matches[0]
+
+    def _get_filename(self, event, exception_idx, frame_idx):
+        exceptions = event.interfaces["exception"].values
+        frame_list = exceptions[int(exception_idx)].stacktrace.frames
+        frame = frame_list[int(frame_idx)]
+        filename = frame.filename
+        return filename
+
+    def _discover_sourcemap_url(self, artifact, filename):
+        file = artifact.file
+        # Code adapted from sentry/lang/javascript/processor.py
+        sourcemap = file.headers.get("Sourcemap", file.headers.get("X-SourceMap"))
+        sourcemap = force_bytes(sourcemap) if sourcemap is not None else None
+
+        if not sourcemap:
+            parsed_body = file.getfile().read().split(b"\n")
+            if len(parsed_body) > 10:
+                possibilities = parsed_body[:5] + parsed_body[-5:]
```

Comment:
@roggenkemper Seems like we lost some comments here. I really wish we could avoid having to replicate this really tricky logic that exits in `sentry/lang/javascript/processor.py`. If we ever change one we have to propagate the changes to the other which is far from ideal because this code is complicated AF. Any ideas how to improve this @kamilogorek?

Reply:
I'd just extract processor function to some sort of util. There's nothing that ties it to that specific file.
There's already a `src/sentry/utils/javascript.py` that would be a perfect place.